File: src/main/java/com/meti/Stream.java
package com.meti;

public interface Stream<T> {
	boolean allMatch(F1<T, Boolean> predicate) throws StreamException;

	default <R> R fold(R initial, F2<R, T, R> folder) throws EndOfStreamException {
		throw new UnsupportedOperationException();
	}

	default Option<T> fold(F2<T, T, T> folder) throws EndOfStreamException {
		throw new UnsupportedOperationException();
	}

	T head() throws StreamException;

	default <R> Stream<R> map(F1<T, R> mapper) {
		throw new UnsupportedOperationException();
	}

	<R> Stream<R> mapE1(F1E1<T, R, ?> mapper);
}

File: src/main/java/com/meti/Stream.scala
package com.meti

trait Stream[T] {
	@throws[StreamException]
	def allMatch(predicate: F1[T, Boolean]): Boolean

	@throws[EndOfStreamException]
	def fold[R](initial: R, folder: F2[R, T, R]): R

	@throws[EndOfStreamException]
	def fold(folder: F2[T, T, T]): Option[T]

	@throws[StreamException]
	def head: T

	@throws[StreamException]
	def headOptionally: Option[T]

	def map[R](mapper: F1[T, R]): Stream[R]

	def mapE1[R](mapper: F1E1[T, R, _]): Stream[R]
}